File: src/main/java/endytkn/randomEvents/baseEvents/groupFight/GroupFightBaseEvent.java
package endytkn.randomEvents.baseEvents.groupFight;

import endytkn.randomEvents.eventObservers.MinecraftEventsObservers;
import endytkn.randomEvents.randomEvent.RandomEvent;
import endytkn.randomEvents.utils.BlockPosUtils;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.effect.MobEffects;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.Mob;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.Vec3;
import net.minecraftforge.event.entity.living.LivingDeathEvent;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

public class GroupFightBaseEvent extends RandomEvent {
    public static final String name = "group_fight";
    protected final Map<String, GroupFight> mobGroupies = new HashMap<>();
    private final boolean removeDrops;

    private final Consumer<LivingDeathEvent> tickDeathConsumer = this::tickDeath;

    public GroupFightBaseEvent(boolean removeDrops) {
        this.removeDrops = removeDrops;
    }

    public void tickDeath(LivingDeathEvent event) {
        Entity entity = event.getEntity();
        CompoundTag nbt = entity.getPersistentData();
        if (this.status != RandomEventStatus.READY || !nbt.contains("REGroupFightName") || !this.isEntityOfEvent(entity)) {
            return;
        }

        String groupName = nbt.getString("REGroupFightName");
        GroupFight group = this.mobGroupies.get(groupName);
        if (group != null) {
            group.killEntity(entity.getUUID(), event.getSource().getEntity() instanceof Player);
        }

        if (this.verifyGroupiesLeft()) {
            this.onGroupiesKilled();
        }
    }

    @Override
    public void onPrepare() {
        this.distanceThreshold = 100;
        spawnMobs();
        MinecraftEventsObservers.livingDeathObserver.remove(this.tickDeathConsumer);
        MinecraftEventsObservers.livingDeathObserver.add(this.tickDeathConsumer);
        super.onPrepare();
    }

    @Override
    public void onReady() {
        super.onReady();
        for (GroupFight group : mobGroupies.values()) {
            for (Mob mob : group.entities.values()) {
                mob.setNoAi(false);
            }
        }
    }

    @Override
    public void onFinishingCanceled() {
        super.onFinishingCanceled();
        for (GroupFight group : mobGroupies.values()) {
            for (Mob mob : group.entities.values()) {
                mob.remove(Entity.RemovalReason.DISCARDED);
            }
        }
    }

    @Override
    public void onFinishing() {
        super.onFinishing();
        MinecraftEventsObservers.livingDeathObserver.remove(this.tickDeathConsumer);
    }

    public void onGroupiesKilled() {
        setEventStatus(RandomEventStatus.FINISHING_SUCCESS);
    }

    private void spawnMobs() {
        for (GroupFight group : mobGroupies.values()) {
            BlockPos groupPos = BlockPosUtils.findRandomSurfaceBlockNearby(this.level, this.targetBlock, 10, 10);
            if (groupPos == null) {
                groupPos = targetBlock;
            }

            for (Mob mob : group.entities.values()) {
                BlockPos mobPos = BlockPosUtils.findRandomSurfaceBlockNearby(this.level, groupPos, 4, 4);
                if (mobPos == null) {
                    mobPos = groupPos;
                }
                mob.setPos(new Vec3(mobPos.getX(), mobPos.getY(), mobPos.getZ()));
                mob.setPersistenceRequired();
                mob.addEffect(new MobEffectInstance(MobEffects.GLOWING, this.timeLimit));
                mob.setNoAi(true);
                CompoundTag nbt = mob.getPersistentData();
                nbt.putString(this.eventTag, this.id.toString());
                nbt.putString("REGroupFightName", group.groupName);
                if (group.groupHates != null) {
                    nbt.putString("REGroupFightHates", group.groupHates);
                    mob.goalSelector.addGoal(1, new GroupFightGoal(mob, Mob.class));
                }
                mob.addAdditionalSaveData(nbt);
                level.addFreshEntity(mob);
            }
        }
    }

    private boolean verifyGroupiesLeft() {
        for (GroupFight group : mobGroupies.values()) {
            if (group.entitiesLeft > 0) {
                return false;
            }
        }
        return true;
    }
}

File: src/main/kotlin/endytkn/randomEvents/baseEvents/GroupFight/GroupFight.baseEvent.kt
package endytkn.randomEvents.baseEvents.GroupFight

import endytkn.randomEvents.baseEvents.InvasionEvent.InvasionWaveStatus
import endytkn.randomEvents.minecraftEventsObservers.MinecraftEventsObservers
import endytkn.randomEvents.randomEvents.RandomEvent
import endytkn.randomEvents.randomEvents.RandomEventStatus
import endytkn.randomEvents.utils.BlockPosUtils
import endytkn.randomEvents.utils.Observer
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.Tag
import net.minecraft.world.effect.MobEffectInstance
import net.minecraft.world.effect.MobEffects
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.Mob
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.Vec3
import net.minecraftforge.event.entity.living.LivingDeathEvent
import kotlin.collections.get

open class GroupFightBaseEvent(val removeDrops: Boolean) : RandomEvent() {
    companion object {
        val name: String = "group_fight"
    }

    protected val mobGroupies: MutableMap<String, GroupFight> = mutableMapOf()

    override fun onPrepare() {
        this.distanceThreshold = 100
        spawnMobs()
        MinecraftEventsObservers.onLivingDeathEventObserver -= ::tickDeath
        MinecraftEventsObservers.onLivingDeathEventObserver += ::tickDeath
        super.onPrepare()
    }

    override fun onReady() {
        super.onReady()
        for (group in mobGroupies.values) {
            for (mob in group.entities.values) {
                mob.isNoAi = false
            }
        }
    }

    override fun onFinishingCanceled() {
        super.onFinishingCanceled()
        for (group in mobGroupies.values) {
            for (mob in group.entities.values) {
                mob.remove(Entity.RemovalReason.DISCARDED)
            }
        }
    }

    override fun onFinishing() {
        super.onFinishing()
        MinecraftEventsObservers.onLivingDeathEventObserver -= ::tickDeath
    }

    open fun onGroupiesKilled() {
        setEventStatus(RandomEventStatus.FINISHING_SUCCESS)
    }

    private fun spawnMobs() {
        for (group in mobGroupies.values) {
            var groupPos = BlockPosUtils.findRandomSurfaceBlockNearby(this.level!!, this.targetBlock!!, 10, 10)
            if (groupPos == null) groupPos = targetBlock

            for (mob in group.entities.values) {
                var mobPos = BlockPosUtils.findRandomSurfaceBlockNearby(this.level!!, groupPos!!, 4, 4)
                if (mobPos == null) mobPos = groupPos
                mob.setPos(Vec3(mobPos.x.toDouble(), mobPos.y.toDouble(), mobPos.z.toDouble()))
                mob.setPersistenceRequired()
                mob.addEffect(MobEffectInstance(MobEffects.GLOWING, this.timeLimit))
                mob.isNoAi = true
                val nbt: CompoundTag = mob.persistentData
                nbt.putString(this.eventTag, this.id.toString())
                nbt.putString("REGroupFightName", group.groupName)
                if (group.groupHates != null) {
                    nbt.putString("REGroupFightHates", group.groupHates)
                    mob.goalSelector.addGoal(1, GroupFightGoal(mob, Mob::class.java))
                }
                mob.addAdditionalSaveData(nbt)
                level!!.addFreshEntity(mob)
            }
        }
    }

    private fun tickDeath(event: LivingDeathEvent) {
        val entity = event.entity
        val nbt: CompoundTag = entity.persistentData
        if (
            status != RandomEventStatus.READY ||
            !nbt.contains("REGroupFightName") ||
            !this.isEntityOfEvent(entity)
        ) return;

        val groupName = nbt.getString("REGroupFightName")
        val group = this.mobGroupies[groupName]

        group!!.killEntity(entity.uuid, event.source.entity is Player)

        if (verifyGroupiesLeft())
            onGroupiesKilled()
    }

    private fun verifyGroupiesLeft(): Boolean {
        mobGroupies.forEach { (_, value) ->
            if (value.entitiesLeft > 0) return false
        }
        return true
    }

}